### DISCUSSION TITLE
HID braille support in NVDA

### DISCUSSION THREAD

**Discussion by Poon-G** [nvaccess/nvda#14877] [Q&A] [answered]
Hi,
I am trying to understand few things in regard to HID support of NVDA. Please help me with below:
1. So with HID, we don't need a separate driver, right? I was testing NVDA HID with orbit20, and noticed orbit20 comes with a default driver: USB composite device which comes under HID layer in device manager. Is this the driver to be used to get HID working with NVDA? Or if not, how to use NVDA with HID device?
2. In NVDA's documentation, it says NVDA is using readFile/WriteFile to read and write data from/to device if connected via HID communication. From ReadFile/WriteFile, how is it converted into braille dots (when in write mode) or text input (when in read mode from device)? Is NVDA using brltty read/write functions after readFile/writeFile HID to convert into the braille format?
3. does composite device connection works with HID NVDA?
4. how multiple languages support (provided via liblouis) is handled by HID? 
5. does blinking cursor support works with HID NVDA?

**Comment by seanbudd** [comment, marked as answer]
1. Exactly, there is no need for a separate driver. That is a core purpose of the HID standard.
1. we use liblouis for braille translation. we do support brltty for older devices if it is installed, however brltty doesn't perform translation. The braille translation implementation should not affect Braille HID device development.
1. I believe these are detected as separate devices
1. The braille translation implementation should not affect Braille HID device development.
1. yes

**Reply by Poon-G** [reply]
Thank you for the reply. Just few follow-ups:
1. So, the driver coming with the device itself is then supposed to work with HID communication?
2. In this I meant, if HID readfile/writefile apis are sufficient to give us required data in the form of braille? OR NVDA is using some conversion functions to convert into braille dots for HID?
4. So liblouis is independent of brltty and it's getting used in HID communication as well?

**Reply by seanbudd** [reply]
1. Yes exactly
1. Yes, these API should be sufficient
1. Yes, brltty is optional and discouraged. Liblouis is what NVDA uses to translate braille within NVDA. It's not relevant to HID communication.
